Elevation mask and basic warm-start

henryhallam opened the pull request:
I accidentally the whole[1] warmstart...
@mookerji how do you like them apples?
Works on bench, ready for review. Will HITL test overnight.
cc @fnoble @cbeighley @gsmcmullin 

     - New setting 'solution.elevation_mask'
     - New tracking channel state field 'elevation'
     - After calc_pvt in solution_thread, calc the el and save it in that field
     - Check 'elevation' against the mask in use_tracking_channel()
     - Drop satellites that are below the mask in manage_track()
     - manage_cal_almanac_scores() expanded into manage_warm_start()
       - Uses ephemeris instead of almanac if we have it
       - Gives sats below el mask a minimum score
       - Narrow doppler search window if we have good position and time.

[1] not really.  Other warmstart stuff includes preseeding the TOW etc.

swiftnav-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://zazu.ci.swiftnav.com/job/piksi_firmware_Pull_Requests/571/
Test PASSed.

swiftnav-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://zazu.ci.swiftnav.com/job/piksi_firmware_Pull_Requests/572/
Test PASSed.

henryhallam commented:
Rebased on master

swiftnav-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://zazu.ci.swiftnav.com/job/piksi_firmware_Pull_Requests/577/
Test PASSed.

mookerji commented:
Great! Do we want to have an elevation mask of 5 deg as the default (as opposed to 0). Also, could it configurable via a setting?

Warm start and elevation mask in the same PR?

mookerji commented on src/manage.c:
```diff
@@ -69,18 +70,20 @@ typedef struct {
 extern ephemeris_t es[32];
 
 static float track_cn0_use_thres = 31.0;
+float elevation_mask = 5.0;
```
nit: comment describing units

henryhallam replied:
ok

mookerji commented on src/manage.c:
```diff
@@ -192,42 +195,69 @@ void manage_acq_setup()
 
-static void manage_calc_almanac_scores(void)
+static u16 manage_warm_start(u8 prn, gps_time_t t,
+                             float *dopp_hint_low, float *dopp_hint_high)
```
nit: docstring

henryhallam replied:
ok

mookerji commented on src/manage.c:
```diff
@@ -192,42 +195,69 @@ void manage_acq_setup()
+      dopp_hint = -GPS_L1_HZ * (vector_dot(3, sat_pos, sat_vel) / GPS_C
+                                + position_solution.clock_bias);
+      /* TODO: Check sign of receiver frequency offset correction */
+      if (time_quality >= TIME_FINE) dopp_uncertainty = DOPP_UNCERT_EPHEM;
```
nit: newline for first branch condition

henryhallam replied:
ok

mookerji commented on src/track.c:
```diff
@@ -141,7 +141,7 @@ float propagate_code_phase(float code_phase, float carrier_freq, u32 n_samples)
  * \param cn0_init           Estimated C/N0 from acquisition
```
nit: update docstring

henryhallam replied:
ok

mookerji commented on src/manage.c:
```diff
@@ -192,42 +195,69 @@ void manage_acq_setup()
+    /* Return the doppler hints and a score proportional to elevation */
+    *dopp_hint_low = dopp_hint - dopp_uncertainty;
+    *dopp_hint_high = dopp_hint + dopp_uncertainty;
+    return SCORE_COLDSTART + SCORE_WARMSTART * el / 90;
```
nit: unit tests!

mookerji replied:
`90` -> `90.`?

henryhallam replied:
unit test: I'll make a ticket to add unit tests for all the management functions.  Could be a good one for the "focused tech debt reduction sprint"

mookerji commented on src/solution.c:
```diff
@@ -416,6 +416,20 @@ static msg_t solution_thread(void *arg)
+      /* Save elevation angles */
+      double az, el;
+      DO_EVERY((u32)soln_freq,
+          for (int i = 0; i < n_ready_tdcp; i++) {
```
Is this something that could be defined as a separate function?

henryhallam replied:
ok

mookerji commented on src/manage.c:
```diff
@@ -192,42 +195,69 @@ void manage_acq_setup()
+      wgsecef2azel(sat_pos, position_solution.pos_ecef, &_, &el_d);
+      el = (float)(el_d) * R2D;
+      if (el < elevation_mask)
+        return 0;
```
Does this return code correspond to an enum already defined elsewhere?

henryhallam replied:
I'll define one

swiftnav-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://zazu.ci.swiftnav.com/job/piksi_firmware_Pull_Requests/590/
Test PASSed.

henryhallam commented:
@mookerji Thanks for the review! I've rebased on master and incorporated your suggestions.
Why this is one PR rather than two:
I originally intended to just write the elevation mask, but with the existing sorta-warm-start logic it would immediately reacquire satellites below the mask as soon as it had dropped them.  So I had to redo that logic to exclude sats known to be below the mask, at which point it also made sense to prioritize sats known to be above the mask.

henryhallam commented:
If no further objections, gonna merge this to master (but not v0.19-branch)

swiftnav-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://zazu.ci.swiftnav.com/job/piksi_firmware_Pull_Requests/601/
Test PASSed.
